polynomial.hpp

```diff
@@ -0,0 +1,220 @@
+                    coefficients[i+1] -= coefficients[i] * p.coefficients[1];
+                }
+            }
+            return *this;
+        }
+        auto operator%=(const polynomial &p) -> polynomial & {
+            if (p.degree == 0 && p.coefficients[0] == T{0}) {
+                throw std::invalid_argument("Division by zero polynomial");
+            }
+            polynomial quotient;
+            polynomial remainder = *this;
+            quotient = *this / p;
+            remainder -= quotient * p;
+            *this = remainder;
+            return *this;
+        }
+      public:
+        auto operator()(T value) {
+        	T result = 0;
+            for (int i = this->degree; i >= 0; --i) {
+                result = result * value + this->coefficients[i];
+            }
+            return result;
+        }
+    };
+    template <typename T>
+    inline auto operator+(polynomial<T> p, polynomial<T> q) {
+        p += q;
+        return p;
+    }
+    template <typename T>
+    inline auto operator-(polynomial<T> p, polynomial<T> q) {
+    	p -= q;
+    	return p;
+    }
+    template <typename T>
+    inline auto operator*(polynomial<T> p, polynomial<T> q) {
+    	p *= q;
+    	return p;
+    }
+    template <typename T>
+    inline auto operator/(polynomial<T> p, polynomial<T> q) {
+    	p /= q;
+    	return p;
+    }
+    template <typename T>
+    inline auto operator%(polynomial<T> p, polynomial<T> q) {
+    	p %= q;
+    	return p;
+    }
+    template <typename T>
+    inline auto divide(polynomial<T> p, polynomial<T> q)
+    ->std::pair<polynomial<T>, polynomial<T>> {
+    	polynomial<T> quotient, remainder;
+        quotient = p / q;
+        remainder = p % q;
+        return std::make_pair(quotient, remainder);
+    }
+    template <typename T>
+    inline auto root_rational_candidates(polynomial<T> p) -> std::vector<T>{
+        size_t P = p.coefficients[p.degree];
+        size_t Q = p.coefficients[0];
+        std::vector<T> factors_of_P{};
+        std::vector<T> factors_of_Q{};
+        std::vector<T> factors{};
+        for(std::size_t i = 1; i < P; i++){
+            if(P % i == 0){
+                factors_of_P.push_back(i);
+            }
+        }
+        for(std::size_t i = 1; i < Q; i++){
+            if(Q % i == 0){
+                factors_of_Q.push_back(i);
+            }
+        }
+        size_t duplicates = 0;
+        for(std::size_t i = 0; i < factors_of_P.size; i++){
+            for(std::size_t j = 0; j < factors_of_Q.size; j++){
+                for(std::size_t k = 0; k < factors.size; k++){
+                    if(factors_of_P[i] / factors_of_Q[j] == factors[k]){
+                        duplicates = 1;
+                        break;
+                    }
+                }
+                if(duplicates == 0){
+                    factors.push_back(factors_of_P[i] / factors_of_Q[j]);
+                }
+                duplicates = 0;
+            }
+        }
```

Comment:
You are allowed to change return type. In general you should return range of fraction<T> if T is integral and range of <T> when T is floating point (in fact you should concentrate on the first case).  To avoid duplicates just change use std::set instead of vectors.